src/plugins/cascade/src/main/js/core/Utils.js

```diff
@@ -9,36 +9,95 @@ define(
   ],
     function (Tools) {
       var getImageClassesForDropdown = function (editor, defaultClassList) {
-        return getClassesForDropdown(editor, defaultClassList, "image");
+        return getClassesForDropdown(editor, defaultClassList, "img");
       };
       var getLinkClassesForDropdown = function (editor, defaultClassList) {
-        return getClassesForDropdown(editor, defaultClassList, "link");
+        return getClassesForDropdown(editor, defaultClassList, "a");
       };
       var getClassesForDropdown = function (editor, defaultClassList, element) {
         var editorSettings = editor.settings;
-        var classList = [];
+        var selectedNode = editor.selection.getNode();
+        var nodeName = selectedNode.nodeName;
+
+        if (nodeName.toLowerCase() !== element) {
+          var elementNode = editor.dom.create(element, {});
+         // editor.selection.setNode(elementNode); //adds node to DOM
+         // selectedNode.append(elementNode); // definitely adds node
+         // var selectedNewNode = editor.selection.select(elementNode); //selecting without inserting/appending causes error
+         // selectedNode = editor.selection.getNode();
+          selectedNode = elementNode; //doesn't seem to work against editor.dom.is for some reason, maybe because it's not an actual node?
+        }
 
+        var classList = [];
         var formatMenuItems = editorSettings.style_formats;
-        Tools.each(formatMenuItems, function (formatMenuItem) {
-          if (formatMenuItem.title === "Custom") {
-            Tools.each(formatMenuItem.items, function (format) {
-              var formatName = format.name;
-              if (editor.formatter.canApply(formatName) && format.classes) {
-                Tools.each(format.classes, function (formatClass) {
-                  if (isInElementClassList(formatClass, element, editorSettings)) {
-                    classList.push(formatClass);
-                  }
-                });
-              }
+        var customMenu = formatMenuItems.filter(function (format) {
+          return format.title === "Custom";
+        });
+
+        if (customMenu[0] === null) {
+          return defaultClassList;
+        }
+
+        Tools.each(customMenu[0].items, function (format) {
+          if ((editor.dom.is(selectedNode, format.selector) || matchesSelectorForNewElement(element, editor.selection.getNode(), format.selector, editor)) && format.classes) {
+            Tools.each(format.classes, function (formatClass) {
+              classList.push(formatClass);
             });
           }
         });
 
         return hasAdvancedFormatMenu(editor) ? classList : defaultClassList;
       };
 
+      // Attempts to checks the selector up until the element since it hasn't been inserted into the editor yet
+      var matchesSelectorForNewElement = function (element, selectedNode, selector, editor) {
```

Comment:
I moved the logic added in the earlier `wysiwyg.js` commits to the plugin side & updated it to so [dom.is](https://www.tiny.cloud/docs/api/tinymce.dom/tinymce.dom.domutils/#is) could semi-work against the selector to help determine if a format would apply to the new element. However, it's definitely imperfect & for more complex selectors, those classes will not get added to the dropdown.

Reply:
We ended up reworking this such that we test if the format is applicable to a node which mimics the functionality of TinyMCE's Format dropdown.